### PR TITLE
Allow cropping non-geos areas

### PR DESCRIPTION
**Pull request by mraspaud**
- [x] Tests added 
 - [x] Tests passed 
 - [x] Passes ``git diff origin/main **/*py | flake8 --diff`` 
 - [ ] Fully documented

### CODE REVIEW

**Comment by mraspaud** [comment]
@ameraner you might want to have a look at this also

**Comment by djhoese** [comment]
If I'm not mistaken this will mean that `reduce_data` in Satpy will now cause a performance hit for all LEO data right? For some reason I thought reduce_data depended on this failing to not have reduce_data be performed on LEO/SwathDefinition data.

**Comment by djhoese** [comment]
Yep, this NotImplementedError handling here:

https://github.com/pytroll/satpy/blob/4f6a7c35ce8a620428f51ff086e3276a73c72c97/satpy/scene.py#L903-L926

File: pyresample/geometry.py [hunk 1]
```diff
@@ -2587,11 +2587,9 @@ def get_area_slices(self, area_to_cover, shape_divisible_by=None):
             return x_slice, y_slice
 
         if not self.is_geostationary:
-            raise NotImplementedError("Source projection must be 'geos' if "
-                                      "source/target projections are not "
-                                      "equal.")
-
-        data_boundary = Boundary(*get_geostationary_bounding_box_in_lonlats(self))
+            data_boundary = self.boundary(frequency=100)
```
Comment (ameraner) [inline pyresample/geometry.py:2590]: what is the rationale for choosing `100` here? (we also hardcode it a bit later https://github.com/pytroll/pyresample/blob/4e245a906bf3aeff072f6b61febdd1df298fb425/pyresample/geometry.py#L2597). I see that the geos case has some more handling of the `frequency` here https://github.com/pytroll/pyresample/blob/4e245a906bf3aeff072f6b61febdd1df298fb425/pyresample/geometry.py#L1559-L1569.

**Comment by djhoese** [comment]
Oh I'm wrong. There is an AreaDefinition check at the top of the method. This is probably good then.

**Comment by ghiggi** [comment]
@mraspaud I guess that if `area_to_cover` is `SwathDefinition`, the method fails. Maybe add a check for that ;) 
Also, if I remember correctly, if an area is within another, I think that the intersection now runs infinitely ...

File: pyresample/geometry.py [hunk 2]
```diff
@@ -2587,11 +2587,9 @@ def get_area_slices(self, area_to_cover, shape_divisible_by=None):
...
+        else:
+            data_boundary = Boundary(*get_geostationary_bounding_box_in_lonlats(self))
```
Comment (ghiggi) [inline pyresample/geometry.py:2592]: If we add `get_geostationary_bounding_box_in_lonlats` in https://github.com/mraspaud/pyresample/blob/4e245a906bf3aeff072f6b61febdd1df298fb425/pyresample/geometry.py#L299 `get_bbox_lonlats` you can simplify all the if_else conditions here and just do self.boundary() with a sensible default ...

Comment (ghiggi) [reply pyresample/geometry.py:2592]: Here below: 

`AreaDefBoundary(area_to_cover, 100)` fails if `area_to_cover` is `SwathDefinition`.
Maybe use `area_to_cover.boundary()` also here in all cases (with sensible frequency defaults)

**Comment by mraspaud** [comment]
> Oh I'm wrong. There is an AreaDefinition check at the top of the method. This is probably good then.

I can't see it, could you point me to it please?

File: pyresample/geometry.py (hunk 2 above)
Comment (mraspaud) [reply pyresample/geometry.py:2592]: I think this simplification is better suited for another PR

**Comment by djhoese** [comment]
Here:

https://github.com/pytroll/pyresample/blob/3fa952985b8b4a62f4407e1cdeef166719523e47/pyresample/geometry.py#L2570-L2571

**Comment by djhoese** [comment]
Hhhmmm maybe I should have merged this before my big config PR. Sorry, but I've caused some conflicts.

**Comment by mraspaud** [comment]
Fixed.

**Comment by mraspaud** [comment]
@djhoese please review my two last commits to see if you agree with the refactoring I made.

This refactoring took me some time because I found a bug/inconsistency in the usage of the `frequency` argument in `AreaDefBoundary` and the `boundary` method of area defintion: in the first case it means the step we use to determine the length of the boundary, in the second case it is the length of the boundary itself. I'm working on a fix now (which will be another PR).

**Comment by ghiggi** [comment]
I remember I noticed such inconsistencies already a while ago. For GEO areas is the total number of vertices, while for swath and other areas is the number of steps on each boundary side right? I personally would **really** like to have this as the number of vertices per boundary side, and allow to specify a tuple (y,x) or (x,y) instead of a single value for all dimensions. 

I also remember somewhat that quite ago, by using the step approach, we were discarding boundary corners values, but I think this was fixed by one of mine or your PR @mraspaud.

**Comment by mraspaud** [comment]
@ghiggi I think both approaches are valid, the usage just depends on what the user needs at that point.

in https://github.com/pytroll/pyresample/pull/526 I have now just replaced `frequency` with `bbox_shape` as a more accurate name. I don't change any functionality or behaviour though, so allowing to pass a tuple for example isn't supported (yet?)

**Comment by mraspaud** [comment]
> which then reaches a PendingDeprecationWarning

Indeed, this is now fixed.

**Comment by djhoese** [comment]
@mraspaud I think I understand the last 2 commits. I assume the main motivation was to:

1. Remove use of AreaDefBoundary
2. Use one method for getting the boundaries for both the source and cover(target) areas.

So if that is true then :+1: from me.

**Pull request merged**